src/tokens/ERC1155B.sol

```diff
@@ -0,0 +1,272 @@
+    //////////////////////////////////////////////////////////////*/
+
+    event TransferSingle(
+        address indexed operator,
+        address indexed from,
+        address indexed to,
+        uint256 id,
+        uint256 amount
+    );
+
+    event TransferBatch(
+        address indexed operator,
+        address indexed from,
+        address indexed to,
+        uint256[] ids,
+        uint256[] amounts
+    );
+
+    event ApprovalForAll(address indexed owner, address indexed operator, bool approved);
+
+    event URI(string value, uint256 indexed id);
+
+    /*//////////////////////////////////////////////////////////////
+                             ERC1155 STORAGE
+    //////////////////////////////////////////////////////////////*/
+
+    mapping(address => mapping(address => bool)) public isApprovedForAll;
+
+    /*//////////////////////////////////////////////////////////////
+                            ERC1155B STORAGE
+    //////////////////////////////////////////////////////////////*/
+
+    mapping(uint256 => address) public ownerOf;
+
+    function balanceOf(address owner, uint256 id) public view returns (uint256 bal) {
+        address idOwner = ownerOf[id];
+
+        assembly {
+            // We avoid branching by using assembly to take
+            // the bool output of eq() and use it as a uint.
+            bal := eq(idOwner, owner)
+        }
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                             METADATA LOGIC
+    //////////////////////////////////////////////////////////////*/
+
+    function uri(uint256 id) public view virtual returns (string memory);
+
+    /*//////////////////////////////////////////////////////////////
+                              ERC1155 LOGIC
+    //////////////////////////////////////////////////////////////*/
+
+    function setApprovalForAll(address operator, bool approved) public virtual {
+        isApprovedForAll[msg.sender][operator] = approved;
+
+        emit ApprovalForAll(msg.sender, operator, approved);
+    }
+
+    function safeTransferFrom(
+        address from,
+        address to,
+        uint256 id,
+        uint256 amount,
+        bytes memory data
+    ) public virtual {
+        require(msg.sender == from || isApprovedForAll[from][msg.sender], "NOT_AUTHORIZED");
+
+        require(from == ownerOf[id], "WRONG_FROM"); // Can only transfer from the owner.
+
+        // Can only transfer 1 with ERC1155B.
+        require(amount == 1, "INVALID_AMOUNT");
+
+        ownerOf[id] = to;
+
+        emit TransferSingle(msg.sender, from, to, id, amount);
+
+        require(
+            to.code.length == 0
+                ? to != address(0)
+                : ERC1155TokenReceiver(to).onERC1155Received(msg.sender, from, id, amount, data) ==
+                    ERC1155TokenReceiver.onERC1155Received.selector,
+            "UNSAFE_RECIPIENT"
+        );
+    }
+
+    function safeBatchTransferFrom(
```

Comment:
👋 I think this, `_batchMint`, and `_batchBurn` can all move the `unchecked` to the loop level